lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -222,47 +234,47 @@ def _from_file(self, root_path: Path, file_path: Path) -> List[dict]:
                 rule_file = yaml.safe_load(rf)
 
             except Exception as e:
-                logger.error("Failed to read alert rules from %s: %s", file_path.name, e)
+                logger.error("Failed to read rules from %s: %s", file_path.name, e)
                 return []
 
-            if _is_official_alert_rule_format(rule_file):
-                alert_groups = rule_file["groups"]
-            elif _is_single_alert_rule_format(rule_file):
-                # convert to list of alert groups
+            if _is_official_rule_format(rule_file):
```

Comment:
This would evaluate as True for both alerting and recording groups. And also for mixed groups.
For single rule format though, it would ignore the other type and emit an error.
We need some spec around this. Add some utests?

E.g.

```text
# GIVEN a folder of alert rules
# WHEN parsed as alert rules
# THEN all alert rules are present in the `parsed`

# WHEN parsed as recording rules
# THEN a warning is emitted for every file
# AND `parsed` is empty

# GIVEN a mixed folder
# WHEN parsed as alert rules
# THEN all alert rules are present in the `parsed`
# AND all recording rules are omitted
```

Reply:
This makes the presumption that we are _not_ going to support "single rule per file recording rules", which is not decided upon. 
 
 No unit tests are needed here unless we come to a consensus that "no, we only want to support the official recording rules format". Otherwise, just a couple of unit tests for "yep, recording rules also pass."
 
 The entire original discussion around "just let users put them in the same path" is orthogonal to differing formats, and "document different paths, different default paths in the constructor, but if some user _really wants_ to jam them all in one place, we'll try to give them best effort". 
 
 It would be a muddled relation data key unless we _actually_ parsed them, and that would involve either some `MixedAlertRules` class, the return sifting through nested keys, or other -- which it's kind of "best effort, but you shouldn't".
 
 Failing that, there doesn't seem to be any real agreement about dropping single-rule-per-file formats, that we should actually validate any of them (that recording rules do not have annotations, that alert rules/recording rules have the right type, and so on, outside of the very small assertion present in `_is_single_rule_format`, which wouldn't have worked with recording rules all in the same bag _anyway_.
 
 In short, this PR/draft was sort-of "put your money where your mouth is" based on what I chatted about for separation.
 
 If we want to have a longer conversation about whether or not we should support single rules or not as a spec document, deprecating old versions, changing databag keys (API break), changing formatting (potential API break), we can, but this PR isn't the right place for that conversation. PR comments just aren't suited for it. 
 
 It will also, inevitably, delay the implementation of Mimir self-monitoring, which seems to rely on some of the recording rules. That's not a blocker, but it's a consideration.
 
 A general sense of "yes, this pattern is fine, let's address any glaring issues and have a follow-up to waffle about what `typing.Whatever` some annotation is", or "no, I think we actually need to have a spec discussion", or "what about this entirely different option?" is what I was hoping for.

Reply:
I'm a bit confused by your response. Afaict,
1. The current code has a bug: different combos have inconsistent outcome
2. We need utests around this

Reply:
The current code does not have a bug. Yet. 

Let me try again:

When we discussed the need for recording rules for Mimir monitoring, there was a discussion about whether we should just go with the solution of mixing them in the same folder/databag, because prom/Mimir allow this. 

I thought we should separate them in the API and databags. After some discussion on mattermost, we proposed different folders in the docs, examples, and constructor, but backend code mostly shared (with a very small child class for each) to give us flexibility in case we need to change, but that we would not explicitly block users who wanted to mix them and set the same path for both, or mixed rules in a folder. 

This is that PR. The inconsistent outcome is consistent with the "best effort support user override" compromise, and the suggested unit tests would forbid/block that. 

This PR could be a concrete/final implementation (after unit tests for recording rules themselves are added), but it was created at least in part for discussion about "this is how my proposed solution would look in real code."

Before raising inconsistencies, we first need to agree on whether this is actually a compromise we can all live with. Then whether or not we actually DO want to forbid mixing, and add logic. It's undefined behavior right now more than a bug, because we have no defined it. 

What are your thoughts on it _overall_? Yes? No? Change some part of the proposed compromise? Agree to formalize some set of behavior?